### PR TITLE
[incubator-kie-issues#69] Time with a timezone is parsed into type Parsed

### LINKED ISSUE #69: Time with a timezone is parsed into type Parsed
Time strings containing timezone, like `@"10:10:10@Australia/Melbourne"` are parsed into a Java type `Parsed` in Drools DMN engine. This type is not supported in the operations in the InfixOpNode, so when it is used in e.g. addition or subtraction, the engine returns null for cases when a result should be returned. 

Investigate and fix this behaviour. Find out if parsing these time strings into `Parsed` type is an accident (uncovered some condition when parsing) or if we should support it in `InfixOpNode` and fix the problem based on the investigation. 

Acceptance criteria: 
- Use cases for addition and subtraction where time with a timezone is used work. 
- New code and these use cases have 100% code coverage.

### PR DESCRIPTION
**Pull request by gitgabrio**
Fixes https://github.com/apache/incubator-kie-issues/issues/69

This PR:
1. creates ZonedOffsetTime decorator to provide both offset and zone information over a time with timezone
2. adapt TimeFunction to return the above, when needed, implementing the missing use-cas (when parsed object contains a zone)
3. add minor required fixes for the new class usage
4. add relevant tests 

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request and downstream checks</b>  
  - Push a new commit to the PR. An empty commit would be enough.

- for a <b>full downstream build</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- for <b>Jenkins PR check only</b>
  - If you are an ASF committer for KIE podling, login to Jenkins (https://ci-builds.apache.org/job/KIE/job/drools/), go to the specific PR job, and click on `Build Now` button.
</details>

### CODE REVIEW

**Review by baldimir — APPROVED** [review]

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/custom/ZonedOffsetTime.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import java.time.temporal.TemporalQuery;
+import java.time.temporal.TemporalUnit;
+import java.util.Objects;
+
+/**
+ * This class is meant as sort-of <b>decorator</b> over <code>OffsetTime</code>, that is a final class.
+ * It is used to provide both offset and zoneid information, replacing the <code>Parsed</code> instance that would be returned otherwise by
+ * {@link  org.kie.dmn.feel.runtime.functions.TimeFunction#invoke(String)}
+ */
+public final class ZonedOffsetTime
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/custom/ZonedOffsetTime.java]: @gitgabrio Could make sense to unit test this new class?

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELDateTimeDurationTest.java [hunk 2]
```diff
@@ -184,6 +192,16 @@ private static Collection<Object[]> data() {
...
 
+                { "@\"P1D\" + @\"2023-10-10T10:31:00@Australia/Melbourne\"", DateTimeFormatter.ISO_DATE_TIME.parse("2023-10-11T10" +
+                                                                                                                             ":31+11:00[Australia/Melbourne]", ZonedDateTime::from), null},
+                { "@\"-P1D\" + @\"2023-10-10T10:31:00@Australia/Melbourne\"", DateTimeFormatter.ISO_DATE_TIME.parse("2023-10-09T10" +
+                                                                                                                             ":31+11:00[Australia/Melbourne]", ZonedDateTime::from), null},
+                { "@\"P1D\" + @\"10:15:00@Australia/Melbourne\"", getCorrectOffsetTime( "10:15", "Australia/Melbourne"), null},
+                { "@\"-P1D\" + @\"10:15:00@Australia/Melbourne\"", getCorrectOffsetTime( "10:15", "Australia/Melbourne"), null},
+                { "@\"PT1H\" + @\"10:15:00@Australia/Melbourne\"", getCorrectOffsetTime( "11:15", "Australia/Melbourne"), null},
+                { "@\"-PT1H\" + @\"10:15:00@Australia/Melbourne\"", getCorrectOffsetTime( "09:15", "Australia/Melbourne"), null},
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/FEELDateTimeDurationTest.java]: @gitgabrio Can you please add some test cases with Substract?

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/ComposingDifferentFunctionsTest.java [hunk 3]
```diff
@@ -101,6 +101,6 @@ void composite5() {
         assertThat(timeOnDateTime.query(TemporalQueries.localTime())).isEqualTo(LocalTime.of(10, 20, 0));
         assertThat(timeOnDateTime.query(TemporalQueries.zone())).isEqualTo(ZoneId.of("Europe/Paris"));
 
-        FunctionTestUtil.assertResult(stringFunction.invoke(timeOnDateTime), "10:20:00@Europe/Paris");
+        FunctionTestUtil.assertResult(stringFunction.invoke(timeOnDateTime), "10:20:00+02:00");
```
Comment (yesamer) [inline kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/ComposingDifferentFunctionsTest.java]: @gitgabrio That means, we previously return `10:20:00@Europe/Paris` to the user and now we always return 10:20:00+02:00, right?

**Review by yesamer — APPROVED** [review]
Approved, just a few minor questions

File: kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/ComposingDifferentFunctionsTest.java (hunk 3 above)
Comment (gitgabrio) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/ComposingDifferentFunctionsTest.java]: @yesamer not really.
This method tests the "toString" of instantiated object, that is an implementation detail.
Users should use the spec-defined methods to access information (e.g. `.timezone`).
This is (implicitly) verified by the fact that no test have been modified in the `FeelDateTimeDurationTest`: does this make sense ?

Comment (yesamer) [reply kie-dmn/kie-dmn-feel/src/test/java/org/kie/dmn/feel/runtime/functions/ComposingDifferentFunctionsTest.java]: @gitgabrio Yes, thanks

**Pull request merged**